pkg/resource/deploy/step_generator.go

```diff
@@ -1521,6 +1521,23 @@ func (sg *stepGenerator) generateStepsFromDiff(
 // calls into continueStepsFromDiff and then validateSteps to continue the work that GenerateSteps would have
 // done with a synchronous diff.
 func (sg *stepGenerator) ContinueStepsFromDiff(event ContinueResourceDiffEvent) ([]Step, error) {
+	// If we're re-entering the step generator we need to handle the case that another step generation that
+	// ran between this event starting and its diff returning has marked this resource deleted and it now
+	// needs to be re-created.
+	urn := event.URN()
+	if sg.deletes[urn] {
+		logging.V(7).Infof("Planner decided to re-create replaced resource '%v' deleted due to dependent DBR", urn)
+
+		// Unmark this resource as deleted, we now know it's being replaced instead.
+		delete(sg.deletes, urn)
+		sg.replaces[urn] = true
```

Comment:
We don't need a lock or anything here do we? As in, there's not some possible race condition where we delete from `deletes` but before we add to `replaces` it gets re-added to `deletes` or some such craziness?